### DISCUSSION TITLE
lf_object, lf_creature

### DISCUSSION THREAD

**Discussion by FrozenEmpire** [ChilliHugger/The-Lords-Of-Midnight#181] [Code Algorithm Questions] [answered]
Hi Chris, todays query :)

I get the impression lf_creature is not only for creatures, but also for other objects placed randomly into locations in the  mxmap::PutThingsOnMap function.

Whereas lf_object signals that one of the unique objects is in the location

Would you say that's right ?

**Comment by FrozenEmpire** [comment]
The auto-seek function in ddr_character::Cmd_WalkForward uses getLocationObject, which only seems to return objects in locations marked with lf_creature. Does that mean it won't auto-find any unique objects marked with lf_object ?

if ( !IsAIControlled() && perform_seek ) {
    // quick fix
    if ( exit_tunnel ) flags.Reset ( cf_tunnel );
    if ( mx->gamemap->getLocationObject(this, Location())!=OB_NONE /*|| ( Location().x==24 && Location().y==58)*/ ) {
        Cmd_Seek();
    }
    if ( exit_tunnel ) flags.Set ( cf_tunnel );
}

**Reply by IcemarkUK** [reply]
That's correct. You never find a unique object by seeking. You are told that one exists in one of the think pages, and you then have to TAKE it as a command.

**Comment by IcemarkUK** [comment, marked as answer]
> Hi Chris, todays query :)
> 
> I get the impression lf_creature is not only for creatures, but also for other objects placed randomly into locations in the mxmap::PutThingsOnMap function.
> 
> Whereas lf_object signals that one of the unique objects is in the location
> 
> Would you say that's right ?

creature / thing / object - has been a naming bane of my life. I've yet managed to refactor the code in anyway where I can be consistent. I think I have it better in the .net rewrite... but still.

lf_creature should be lf_thing and relates to
```
        enum mxthing_t {
            // COMMON
            OB_NONE=0,
            OB_WOLVES,
            OB_DRAGONS,
            OB_ICETROLLS,
            OB_SKULKRIN,
            OB_WILDHORSES,
            OB_SHELTER,
            OB_GUIDANCE,
            // LOM
            OB_SHADOWSOFDEATH,
            OB_WATERSOFLIFE,
            OB_HANDOFDARK,
            OB_CUPOFDREAMS,
            OB_WOLFSLAYER,
            OB_DRAGONSLAYER,
            OB_ICECROWN,
            OB_MOONRING,
            // DDR
            OB_CLAWS,
            OB_FLAMES,
            OB_THORNS,
            OB_BLOOD,
            OB_LANGUOR,
            OB_SPRINGS,
        };
```
remembering that these are TME combined values and not original LOM/DDR - and yes lf_object means a unique object.

But as you can see the mxthing_t enum also actually relates to an object. Just the initial ones that are not unique!! The Crown of Varenand / OB_CROWN_VARENAND in my code base being ID 22 - where OB_SPRNGS is 21.

**Reply by FrozenEmpire** [reply]
Not unique.... apart from OB_ICECROWN and OB_MOONRING which shouldn't be part of mxthing_t really ?

**Reply by IcemarkUK** [reply]
Yes,  OB_ICECROWN and OB_MOONRING were the only unique objects in LoM and ideally they should move to the DDR mechanic of having unique objects.

if it wasn't for the fact that it would need a save game conversion, then I would change mxthing_t to have the prefix TH_ and move ICECROWN and MOONRING out. lf_creature would then become lf_thing and objects would be purely the uniques.

I might spin up a branch and see what that looks like...

**Comment by IcemarkUK** [comment]
@FrozenEmpire I've been looking at this tentatively and a number of issues come up that need consideration with backward compatibility (for me).  [PR for spike](https://github.com/ChilliHugger/The-Lords-Of-Midnight/pull/182)

Talking about it here for the benefit of "rubber ducking"...

Map. object on map is current 9 bits - 511 which is too big for map things. Dropping it down to 6 bits gives a range of 63 which is probably big enough for how they are being used. This creates a problem with current objects ie: OB_MOONRING / OB_ICECROWN and all DDR objects as they flow in to this range.

This all means that the data needs to be remapped internally. Ideally it needs an mxthing* as well as mxobject* with mxobject only being the uniques.

**Reply by FrozenEmpire** [reply]
Hi Chris,

I can see how easy it is to get knotted up with this - was also thinking how best to simplify it - although in my case I’m not constrained by data bits :)

If your aim is backward compatibility then how about this:

Keep the single type OB_ which includes all objects as now. Forget about separating out TH_. Get rid of the creature flag.

Each object already has properties to know whether it’s unique, can be seen etc. why not use these properties rather than separating out uniques etc.

For example, instead of testing whether a creature/thing is at the location, just need to check if any non unique object is there.

And when testing if there is anything that can be seen, just test for unique objects that can be seen.

Maybe I’m missing some circumstances where this will break, but seems viable?

**Reply by IcemarkUK** [reply]
From here I am going to say THING for creature and OBJECT for the unique objects.

A THING and a OBJECT can be in the same location. This is the difference between the two games. In LoM there was just a map thing  DDR has a map THING then all the objects are a lookup. So lf_thing flag (lf_creature previously) means that the location has a thing and it's calculated algorithmically. object flag means there is a lookup for one of the 128 objects.

I purposely tried to get rid of the thing algorithm and have it stored on the map. That brought LoM and DDR in line, but allowed for expansion - however the algorithm is still used every night in the ddr scenario to work out what thing needs to be added back to the map. Again there is The Citadel to think about.

So lf_thing means the map thing field for a location is set, and lf_object means there is also a unique object that can be carried.
Technically lf_thing flag could be removed because the object variable would be OB_NONE or a valid value.

LoM has 4 objects that are problematic. Sword Dragonslayer and Wolveslayer are non-unique carry-able objects, and Icecrown and Morning as unique carry-able objects. So they fall in to being THINGS because there was no concept of unique objects. When I joined the two games I ended up with mxobject which was the combination  which could be carried. those objects have a flag to say they can be carried. 

DDR has 128 objects that can't be stored in the map variable because they can be in a location with a THING.

So the map object variable has a whopping 511 values which are unlikely to ever be used. This is just me messing up the combined data all those years ago.

So my proposal that I am looking at is to cap the number of things in a location to 63. Leave the object database as is, containing things and objects. Add a new field to the object which is what 'thing' it relates to on the map (if it can be on the map) and therefore mxthing_t is the enum of just the mappable things.

I then have a remap table for the 64 'things' that gives you their proper object. So then internally when something is found on the map, it remaps to the property object. 

What this means is that in LoM if you seek and find the Icecrown or Moonring it is removed from the map. Because they are also marked as unique their carriedby variable is set. If they are then later dropped (due to death) then lf_object is set, the location of the object is stored (just like the DDR objects) but the map thing field is not set, so the location does not get flagged with lf_thing. Because from this point on it is handled by the stand unique object system.

If you seek and find a sword then, lf_thing is reset, the map thing field is cleared, the sword object is then grabbed from the object and stored in the characters 'carried' variable. However as the object is not unique it does not have its 'carriedby' field set. If that sword is later dropped (by picking up another one) the map's lf_thing is set and the thing field is set to the 'thing' field (1-63).

Essentially, the four problematic THINGS in LoM become a marker so that when you seek the engine says - oh, these are special, you've found an object, gets the object from the object table and returns it instead.

I know that might sound long winded and complicated... :)

**Reply by FrozenEmpire** [reply]
Well, it aced my very much over simplified suggestion !

Now I really understand what you're going through :)

**Reply by IcemarkUK** [reply]
Perfectly example of getting the architecture wrong way back...

**Comment by FrozenEmpire** [comment]
An additional related question

mxscenario::PlaceObjectsOnMap() places the unique objects around the map at scenario creation, avoiding locations which already have an object in them...

mxmap::PutThingsOnMap() is called during each Night routine, and refreshes the non-unique objects in buildings and terrains... however there is no check to see if there is already a unique object in the location, so unique objects could seemingly get overwritten here...

Would you agree ?

**Reply by IcemarkUK** [reply]
So.. calculateNameSeed(x,y) will always return the same value for the same given x,y

The following terrains (PLAINS,MOUNTAIN,FOREST,HILLS) will randomly have their thing put back in their location based on ( r = mxrandom(3) ). It will always be the same thing.

All the others will have their thing put back every night.

Unique objects are not stored on the map. lf_object means - there is an object here, go look it up.

so things and objects co-exists at a location.

**Reply by FrozenEmpire** [reply]
Got it. The bit I was missing was that unique objects are only flagged but not stored.

**Comment by IcemarkUK** [comment]
One interesting thing I've just found while doing this... I have an existing bug... I seem to have lost the ability for Luxor to drop the moonring! Not sure if I just missed it a long time ago, or I broke it sometime... Which means if he dies, there is no way for Morkin to pick it up...

**Reply by IcemarkUK** [reply]
Phew! Luckily something I just broke during this spike - and not part of the actual codebase!

**Comment by IcemarkUK** [comment]
@FrozenEmpire So I am at the end of the spike. Ultimately I need to decide if I want to take the risk of this refactor. Really I should write a number of unit tests. In the long run the code will be slightly better for it. But is it worth the upheaval...?

**Reply by FrozenEmpire** [reply]
I think I would be inclined to merge as part of the next significant update to the codebase, ie The Citadel ;)

**Comment by FrozenEmpire** [comment]
Just to confirm - it seems a character can carry both a THING, and an OBJECT at the same time, right ?

**Reply by IcemarkUK** [reply]
No. In TME they are interchangeable so in LOM Luxor can’t carry the moon ring (thing/object) and the sword wolfslayer (thing) at the same time. In DDR there are no things that can be carried just objects.

**Reply by FrozenEmpire** [reply]
Ah yes.. of course. No carriable THINGS in DDR !
